kernel/comps/virtio/src/queue.rs

```diff
@@ -32,151 +33,135 @@ pub enum QueueError {
 
 /// The mechanism for bulk data transport on virtio devices.
 ///
-/// Each device can have zero or more virtqueues.
+/// A device can have zero or several virtqueues.
+///
+/// The const parameter `SIZE` specifies the size of the queue,
+/// correlating to the number of descriptors, and the number of slots
+/// in both the available and used rings.
 #[derive(Debug)]
-pub struct VirtQueue {
-    /// Descriptor table
-    descs: Vec<SafePtr<Descriptor, DmaCoherent>>,
-    /// Available ring
-    avail: SafePtr<AvailRing, DmaCoherent>,
-    /// Used ring
-    used: SafePtr<UsedRing, DmaCoherent>,
-    /// point to notify address
+pub struct VirtQueue<const SIZE: usize> {
```

Comment:
The SIZE of VirtQueue should not be a constant.

The actually queue size should be determined after driver negotiates with the device at run time. It cannot be determined at compile time.

Reply:
The reason I choose to define a constant generic is to make the `VirtQueue` typesafe. This is becasue the inner  `UsedRingBuf` and `AvailRingBuf` must  be a `Pod` type, and the size of the ring is determined by the `VirtQueue`. 

currently, the `AvailRing` and `UsedRing` use `64` as its size, if the `VirtQueue` uses a bigger size, this will lead to the occurrence of unknown behaviors.

Reply:
I have reservation about having the queue size statically set, too. See my comment "The driver cannot unilaterally determine the queue size at compile time" below .